src/mscorlib/src/System/Threading/LazyInitializer.cs

```diff
@@ -217,25 +219,47 @@ public static T EnsureInitialized<T>(ref T target, ref bool initialized, ref obj
         /// <returns>The initialized object.</returns>
         private static T EnsureInitializedCore<T>(ref T target, ref bool initialized, ref object syncLock, Func<T> valueFactory)
         {
-            // Lazily initialize the lock if necessary.
             object slock = syncLock;
-            if (slock == null)
+
+            // Lazily initialize the lock if necessary and,  then double check if initialization is still required.
+            lock (EnsureLockInitialized(ref slock))
```

Comment:
Yes, agreed. This is because of oversight. My test cases ran successfully because I added one for non-null syncLock to the new LazyInitializer.EnsureInitialized API and was happy with its success. I will add another test case for scenario that the user sends in a null syncLock which would have set off the race condition.